### ISSUE TITLE
Newlines at the end of messages at the `aws_kinesis_firehose` sink

### ISSUE THREAD

**Issue by MOZGIII** [vectordotdev/vector#3849]
See https://ptb.discordapp.com/channels/742820443487993987/746070591097798688/751515481239847002.

User had an issue with `aws_kinesis_firehose` in particular, but it may be an issue with our sinks in general.

We remove trailing newlines from Event `message` strings after parsing the input lines, as they lose their meaning after being parsed. This is a good, semantically correct way of operating with log lines. - the rationale is the process of parsing transforms a stream framed bytes (POSIX notion of lines) into an in-memory representation of a line that is defined differently (the `Bytes` type), thus eliminating the need to keep the pre-parsing framing details of each logline.

At the same time, sinks pass might pass the data to somewhere where the data is effectively just concatenated and written to a file - like `aws_kinesis_firehose` sink that simply dumps the contents to S3 at the receiving end. In this case, the lack of `\n` is becoming a problem.

I view this as a limitation of the encoding. Currently, we only have a `text` encoding in that sink - that simply passes each event's `message` out - but for a situation like the above, it would be meaningful to also have something like `lines` ("line" being a sequence of bytes terminated with `\n`, according to POSIX definition of a line), or `newline_terminated_text` for better clarity.

The same idea could be applied to other sinks, enabling more use cases with less configuration.

Of course, it's possible to alter the message in a transform before passing it to sink - and simply add a trailing newline manually. However, that would be less efficient than implementing it at the codec level, and it's also a manual work for something that can be considered a generic case.

**Comment by binarylogic** [comment]
> Currently, we only have a text encoding in that sink

We also offer `json` encoding.

It's still not entirely clear to me what the problem is. There's a lot of talk about things outside of the core problem that make this hard to follow. Maybe strip down the text and try to demonstrate what you mean?

**Comment by MOZGIII** [comment]
I think it's easy to demonstrate with an example:

input: `a\nb\nc`
source parses into: `a` `b` `c`
sink encodes into : `a` `b` `c`
s3 writes: `abc`

We could instead:

input: `a\nb\nc`
source parses into: `a` `b` `c`
sink encodes into : `a\n` `b'n` `c\n`
s3 writes: `a\nb\nc`

**Comment by jszwedko** [comment]
To add some more detail, it looks like it is Firehose itself that is concatenating the messages when delivering downstream:

> Kinesis Data Firehose buffers records before delivering them to the destination. To disambiguate the data blobs at the destination, a common solution is to use delimiters in the data, such as a newline (\n) or some other character unique within the data. This allows the consumer application to parse individual data items when reading the data from the destination. 

https://docs.aws.amazon.com/firehose/latest/APIReference/API_PutRecordBatch.html

For JSON I think this would probably be fine, but even then people might prefer to suffix a newline.

I think it'd be nice to be able to specify the record delimiter (for example some people might prefer `\0` terminated). Would we preserve the escaping of any newlines in the event itself?

I agree that this seems like the common case for Kinesis Firehose so we should aim to support it without additional configuration rather than requiring users to transform the records.

**Comment by jszwedko** [comment]
Closing in-lieu of https://github.com/vectordotdev/vector/issues/8487 which has more details.
